File: dashboard_app/city_catalog.js
class UnsupportedCityError extends Error {
  constructor(cityKey) {
    super(`City "${cityKey}" is not supported.`);
    this.name = "UnsupportedCityError";
  }
}

const CITIES = [
  { key: "toronto", label: "Toronto" },
  { key: "ottawa", label: "Ottawa" },
  { key: "vancouver", label: "Vancouver" },
  { key: "montreal", label: "Montreal" },
  { key: "calgary", label: "Calgary" },
];

function listCities() {
  return CITIES.map((city) => ({ ...city }));
}

function getCity(cityKey) {
  const normalizedKey = String(cityKey).trim().toLowerCase();
  const city = CITIES.find((entry) => entry.key === normalizedKey);

  if (!city) {
    throw new UnsupportedCityError(cityKey);
  }

  return { ...city };
}

module.exports = {
  UnsupportedCityError,
  getCity,
  listCities,
};

File: weather_etl/city_catalog.py
SUPPORTED_CITIES = {
    "toronto": {
        "name": "Toronto",
        "latitude": 43.6532,
        "longitude": -79.3832,
        "timezone": "America/Toronto",
    },
    "ottawa": {
        "name": "Ottawa",
        "latitude": 45.4215,
        "longitude": -75.6972,
        "timezone": "America/Toronto",
    },
    "vancouver": {
        "name": "Vancouver",
        "latitude": 49.2827,
        "longitude": -123.1207,
        "timezone": "America/Vancouver",
    },
    "montreal": {
        "name": "Montreal",
        "latitude": 45.5017,
        "longitude": -73.5673,
        "timezone": "America/Toronto",
    },
    "calgary": {
        "name": "Calgary",
        "latitude": 51.0447,
        "longitude": -114.0719,
        "timezone": "America/Edmonton",
    },
}


def get_city(city_key: str) -> dict:
    normalized_key = city_key.lower()
    if normalized_key not in SUPPORTED_CITIES:
        raise ValueError(f"Unsupported city: {city_key}")

    return {"id": normalized_key, **SUPPORTED_CITIES[normalized_key]}
